File: client/c/client/src/client/ServiceProxy.cpp
/*
 *  Copyright Beijing 58 Information Technology Co.,Ltd.
 *
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
/*
 * ServiceProxy.cpp
   *
 * Created on: 2011-7-12
 * Author: Service Platform Architecture Team (spat@58.com)
 */

#include "ServiceProxy.h"
#include "Dispatcher.h"
#include "GaeaClientConfig.h"
#include "Parameter.h"
#include "../protocol/SdpStruct.h"
#include "../serialize/serializeList.h"
#include "../serialize/serializer.h"
#include "../serialize/derializer.h"
#include "../serialize/strHelper.h"
#include "GaeaConst.h"
#include "../protocol/Protocol.h"
#include "Log.h"
#include <stdexcept>
#include <string>
#include <pthread.h>
#include <errno.h>
#include <signal.h>
namespace gaea {
pthread_mutex_t ServiceProxy::mutex = PTHREAD_MUTEX_INITIALIZER;
std::map<std::string, ServiceProxy*> ServiceProxy::proxys;
volatile int ServiceProxy::sessionId = 1;
ServiceProxy::ServiceProxy(std::string &serviceName) {
	config = GaeaClientConfig::getConfig(serviceName);
	dispatcher = new Dispatcher(config);
}
ServiceProxy* ServiceProxy::getProxy(std::string &serviceName) {
	std::map<std::string, ServiceProxy*>::iterator it = proxys.find(serviceName);
	ServiceProxy *proxy = NULL;
	if (it == proxys.end()) {
		pthread_mutex_lock(&mutex);
		it = proxys.find(serviceName);
		if (it == proxys.end()) {
			proxy = new ServiceProxy(serviceName);
			proxys.insert(std::pair<std::string, ServiceProxy*>(serviceName, proxy));
		} else {
			proxy = it->second;
		}
		pthread_mutex_unlock(&mutex);
	} else {
		proxy = it->second;
	}
	return proxy;
}
ServiceProxy::~ServiceProxy() {
	delete dispatcher;
}
InvokeResult* ServiceProxy::invoke(char *lookup, char *methodName, Parameter **paras, int parasLen) {
	serialize_list *paraList;
	paraList = NULL;
	InvokeResult *is = NULL;
	Protocol *receiveP;
	KeyValuePair kvp[parasLen];
	std::string s = "RpParameter";
	for (int i = parasLen - 1; i >= 0; --i) {
		kvp[i].key = paras[i]->getSimpleName();
		kvp[i].value = paras[i]->getValue();
		kvp[i].valueTypeId = paras[i]->getTypeId();
		paraList = list_cons(&(kvp[i]), GetTypeId((char*) s.c_str()), paraList);
	}
	RequestProtocol requestProtocol;
	requestProtocol.lookup = lookup;
	requestProtocol.methodName = methodName;
	requestProtocol.paraList = paraList;
	std::string sdpEntityType = "RequestProtocol";
	int sid = createSessionId();
	Protocol sendP(sid, config->getServiceId(), Request, UnCompress, GAEABinary, Java, &requestProtocol, sdpEntityType.c_str());
	for (int i = 0; i < 3; ++i) {
		Server *server = dispatcher->getServer();
		if (server == NULL) {
			errno = -3;
			list_free(paraList);
			throw std::runtime_error("cannot get server");
		}
		try {
			receiveP = server->request(sendP);
		} catch (std::exception &e) {
			gaeaLog(GAEA_WARNING,"server request error:hostName:%s   ,error msg:%s\n",server->getName(), e.what());
			continue;
		}
		if (!receiveP) {
			continue;
		} else if (receiveP->getSdpType() == Response) {
			ResponseProtocol *rp = (ResponseProtocol*) receiveP->getSdpEntity();
			is = new InvokeResult(rp->result, rp->outpara);
			free(rp);
			delete receiveP;
			break;
		} else if (receiveP->getSdpType() == Exception) {
			ExceptionProtocol *ep = (ExceptionProtocol*) receiveP->getSdpEntity();
			std::string errorMsg(ep->ErrorMsg);
			DerializeFree((char*) "ExceptionProtocol", ep);
			free(receiveP);
			errno = -2;
			gaeaLog(GAEA_WARNING, errorMsg.c_str());
			break;
		} else if (receiveP->getSdpType() == RebootException) {
			ResetProtocol *rp = (ResetProtocol*) receiveP->getSdpEntity();
			DerializeFree((char*) "ResetProtocol", rp);
			free(receiveP);
			server->markAsReboot();
			errno = -2;
			printf("%d\n", receiveP->getSessionID());
			gaeaLog(GAEA_WARNING, "%s server has reboot,system will change normal server!\n", server->getName());
		} else {
			errno = -2;
			gaeaLog(GAEA_WARNING, ("userdatatype error!"));
		}
	}
	list_free(paraList);
	if (!is) {
		throw std::runtime_error("operation failed!\n");
	}
	return is;
}
int ServiceProxy::createSessionId() {
	int sid = 0;
	pthread_mutex_lock(&mutex);
	sessionId++;
	if (sessionId > MAX_SESSIONID) {
		sessionId = 1;
	}
	sid = sessionId;
	pthread_mutex_unlock(&mutex);
	return sid;
}
InvokeResult::InvokeResult(void *result, array *outPara) {
	this->result = result;
	this->outPara = outPara;
}
InvokeResult::~InvokeResult() {
}

array* InvokeResult::getOutPara() const {
	return outPara;
}

void *InvokeResult::getResult() const {
	return result;
}
}

File: client/java/client/src/com/bj58/spat/gaea/client/proxy/ServiceProxy.java
/*
 *  Copyright Beijing 58 Information Technology Co.,Ltd.
 *
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package com.bj58.spat.gaea.client.proxy;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;

import com.bj58.spat.gaea.client.GaeaConst;
import com.bj58.spat.gaea.client.configuration.ServiceConfig;
import com.bj58.spat.gaea.client.loadbalance.Dispatcher;
import com.bj58.spat.gaea.client.loadbalance.Server;
import com.bj58.spat.gaea.client.proxy.builder.Parameter;
import com.bj58.spat.gaea.client.utility.logger.ILog;
import com.bj58.spat.gaea.client.utility.logger.LogFactory;
import com.bj58.spat.gaea.protocol.exception.RebootException;
import com.bj58.spat.gaea.protocol.exception.ThrowErrorHelper;
import com.bj58.spat.gaea.protocol.exception.TimeoutException;
import com.bj58.spat.gaea.protocol.sdp.ExceptionProtocol;
import com.bj58.spat.gaea.protocol.sdp.HandclaspProtocol;
import com.bj58.spat.gaea.protocol.sdp.RequestProtocol;
import com.bj58.spat.gaea.protocol.sdp.ResponseProtocol;
import com.bj58.spat.gaea.protocol.sfp.enumeration.CompressType;
import com.bj58.spat.gaea.protocol.sfp.enumeration.PlatformType;
import com.bj58.spat.gaea.protocol.sfp.enumeration.SDPType;
import com.bj58.spat.gaea.protocol.sfp.v1.Protocol;
import com.bj58.spat.gaea.protocol.utility.KeyValuePair;

/**
 * ServiceProxy
 *
 * @author Service Platform Architecture Team (spat@58.com)
 */
public class ServiceProxy {

    private Dispatcher dispatcher;
    private ServiceConfig config;
    private int sessionId = 1;
    private int requestTime = 0;//超时重连次数
    private int ioreconnect = 0;//IO服务切换次数
    private int count = 0; //根据requestTime和ioreconnect确定
    private static final ILog logger = LogFactory.getLogger(ServiceProxy.class);
    private static final Object locker = new Object();
    private static final HashMap<String, ServiceProxy> Proxys = new HashMap<String, ServiceProxy>();

    private ServiceProxy(String serviceName) throws Exception {
        config = ServiceConfig.GetConfig(serviceName);
        dispatcher = new Dispatcher(config);
        
        requestTime = config.getSocketPool().getReconnectTime();
    	int serverCount = 1;
    	if(dispatcher.GetAllServer() != null && dispatcher.GetAllServer().size() > 0){
    		serverCount = dispatcher.GetAllServer().size();
    	}
    	
    	ioreconnect = serverCount - 1;
    	
    //	count = max {ioreconnect, requestTime}
    	count = requestTime;
    	
    	if(ioreconnect > requestTime){
    		count = ioreconnect;
    	}
    }
    
    private void destroy() {
    	List<Server> serverList = dispatcher.GetAllServer();
		if(serverList != null) {
			for(Server server : serverList) {
				try {
					server.getScoketpool().destroy();
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
		}
    }

    public static ServiceProxy getProxy(String serviceName) throws Exception {
        ServiceProxy p = Proxys.get(serviceName.toLowerCase());
        if (p == null) {
	        synchronized(locker) {
	        	p = Proxys.get(serviceName.toLowerCase());
	        	if (p == null) {
	        		p = new ServiceProxy(serviceName);
	        		Proxys.put(serviceName.toLowerCase(), p);
	        	}
	        }
        }
        return p;
    }

    public InvokeResult invoke(Parameter returnType, String typeName, String methodName, Parameter[] paras) throws Exception, Throwable {
        long watcher = System.currentTimeMillis();
        List<KeyValuePair> listPara = new ArrayList<KeyValuePair>();
        for (Parameter p : paras) {
            listPara.add(new KeyValuePair(p.getSimpleName(), p.getValue()));
        }
        RequestProtocol requestProtocol = new RequestProtocol(typeName, methodName, listPara);
        Protocol sendP = new Protocol(createSessionId(), 
        		(byte) config.getServiceid(),
        		SDPType.Request,
        		CompressType.UnCompress,
        		config.getProtocol().getSerializerType(),
        		PlatformType.Java,
        		requestProtocol);
        
        Protocol receiveP = null;
        Server server = null;
        
        for(int i = 0; i <= count; i++){
        	server = dispatcher.GetServer();
            if (server == null) {
                logger.error("cannot get server");
                throw new Exception("cannot get server");
            }
            try{
            	receiveP = server.request(sendP);
            	break;
            } catch(IOException io){
            	if(count == 0 || i == ioreconnect){
            		throw io;
            	}
            	if(i < count && i < ioreconnect) {
            		logger.error(server.getName()+" server has IOException,system will change normal server!");
            		continue;
            	}
            } catch(RebootException rb){
            	this.createReboot(server);
            	if(count == 0 || i == ioreconnect){
            		throw new IOException("connect fail!");
            	}
            	if(i < count && i < ioreconnect) {
            		logger.error(server.getName()+" server has reboot,system will change normal server!");
            		continue;
            	}
            }catch(TimeoutException te){
            	if(count == 0 || i == requestTime){
            		throw new TimeoutException("Receive data timeout or error!");
            	}
            	if(i < count && i < requestTime) {
            		logger.error(server.getName()+" server has TimeoutException,system will change normal server!");
            		continue;
            	}
            } catch (Throwable ex){
            	logger.error("invoke other Exception", ex);
            	throw ex;
            }
    	}
        
        if(receiveP == null){
        	throw new Exception("userdatatype error!");
        }
        
        if (receiveP.getSDPType() == SDPType.Response) {
            ResponseProtocol rp = (ResponseProtocol)receiveP.getSdpEntity();
            logger.debug("invoke time:" + (System.currentTimeMillis() - watcher) + "ms");
            return new InvokeResult(rp.getResult(), rp.getOutpara());
        } else if(receiveP.getSDPType() == SDPType.Reset){ //服务重启
        	logger.info(server.getName()+" server is reboot,system will change normal server!");
        	this.createReboot(server);
        	return invoke(returnType, typeName, methodName, paras);
        }else if (receiveP.getSDPType() == SDPType.Exception) {
            ExceptionProtocol ep = (ExceptionProtocol)receiveP.getSdpEntity();
            throw ThrowErrorHelper.throwServiceError(ep.getErrorCode(), ep.getErrorMsg());
        } else {
            throw new Exception("userdatatype error!");
        }
    }
    
    /**
     * 设置当前重启服务
     * @param server
     * @throws Throwable 
     * @throws Exception 
     */
    private void createReboot(Server server) throws Exception, Throwable{
    	server.createReboot();
    }
    
    /**
     * 权限协议
     * @param data 
     * @return Protocol
     * @throws Exception 
     */
    public Protocol createProtocol(HandclaspProtocol hp) throws Exception{
    	Protocol sendRightsProtocol = new Protocol(createSessionId(), 
        		(byte) config.getServiceid(),
        		SDPType.Request,
        		CompressType.UnCompress,
        		config.getProtocol().getSerializerType(),
        		PlatformType.Java,
        		hp);
    	return sendRightsProtocol;
    }
    
    /**
     *  get Server info
     * @param name Server name
     * @return if Server exist return Server info else return empty
     */
    public String getServer(String name) {
        Server server = dispatcher.GetServer(name);
        if (server == null) {
            return "";
        }
        return server.toString();
    }
    
    public static void destroyAll() {
    	Collection<ServiceProxy> spList = Proxys.values();
		if(spList != null) {
			for(ServiceProxy sp : spList) {
				sp.destroy();
			}
		}
    }
    

    private int createSessionId() {
        synchronized (this) {
            if (sessionId > GaeaConst.MAX_SESSIONID) {
                sessionId = 1;
            }
            return sessionId++;
        }
    }
    
    /**
     * invoke result
     * <typeparam name="T">result data type</typeparam>
     */
    public class InvokeResult<T> {

        InvokeResult(Object result, Object[] outPara) {
            Result = (T) result;
            OutPara = outPara;
        }
        private T Result;
        private Object[] OutPara;

        public Object[] getOutPara() {
            return OutPara;
        }

        public void setOutPara(Object[] OutPara) {
            this.OutPara = OutPara;
        }

        public T getResult() {
            return Result;
        }

        public void setResult(T Result) {
            this.Result = Result;
        }
    }
}
